### PR TITLE
Add optional serde support for task type

### PR DESCRIPTION
**Pull request by matthiasbeyer**
Add optional serde support for the `Task` type.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -18,5 +18,18 @@ regex = "^0.2"
 version = "^0.0"
 optional = true
 
+[dependencies.serde]
+version = "1"
```
Comment (sanpii) [inline Cargo.toml]: I don’t know how `1` works when we run `cargo update`. I prefer using `^1.0`.

File: Cargo.toml [hunk 2]
```diff
@@ -18,5 +18,18 @@ regex = "^0.2"
...
+optional = true
+
+[dependencies.serde_derive]
+version = "1"
```
Comment (sanpii) [inline Cargo.toml]: I don’t know how `1` works when we run `cargo update`. I prefer using `^1.0`.

File: Cargo.toml [hunk 3]
```diff
@@ -18,5 +18,18 @@ regex = "^0.2"
...
+optional = true
+
 [badges]
 travis-ci = { repository = "sanpii/todo-txt", branch = "master" }
+
+[features]
+default       = []
+serde-support = ["serde", "serde_derive", "chrono/serde"]
```
Comment (sanpii) [inline Cargo.toml]: What is `chrono/serde`?

File: src/task.rs [hunk 4]
```diff
@@ -1,17 +1,28 @@
...
+    #[cfg_attr(feature = "serde-support", serde(skip_serializing_if = "Option::is_none"))]
     pub threshold_date: Option<::Date>,
+
+    #[cfg_attr(feature = "serde-support", serde(skip_serializing_if = "Option::is_none"))]
     pub due_date: Option<::Date>,
     pub contexts: Vec<String>,
     pub projects: Vec<String>,
     pub hashtags: Vec<String>,
+
+    #[cfg_attr(feature = "serde-support", serde(skip_serializing_if = "BTreeMap::is_empty"))]
```
Comment (sanpii) [inline src/task.rs]: I don’t like the idea to remove empty fields.

On the client side (typically, a javascript front end) you need a special condition to handle the event of missing field. If the field is always available, you can do a for loop on an empty array.

It’s almost the same problem with removing options field. Using a ternary operator is more readable than an `if` condition with `undefined`.

File: Cargo.toml (hunk 3 above)
Comment (matthiasbeyer) [reply Cargo.toml]: This enables the "serde" feature in the "chrono" dependency when "serde-support" is enabled.

**Comment by matthiasbeyer** [comment]
I'm working on patches for the annotated things! :heart:

**Comment by matthiasbeyer** [comment]
I'm adding more tests right now, please hang on.

**Comment by matthiasbeyer** [comment]
Now I guess all the things are fixed.

I will cleanup this PR (remove the fixup commits) before merging, of course. :+1:

File: Cargo.toml [hunk 5]
```diff
@@ -18,5 +18,28 @@ regex = "^0.2"
...
+[dependencies.serde]
+version = "^1"
+optional = true
+
+[dependencies.serde_derive]
+version = "^1"
+optional = true
+
+[dev-dependencies]
+serde        = "^1"
```
Comment (sanpii) [inline Cargo.toml]: `serde` and `serde_derive` are already a dependencies, it’s not necessary to duplicated them here.

File: src/task.rs [hunk 6]
```diff
@@ -84,3 +85,85 @@ impl ::std::fmt::Display for Task
         Ok(())
     }
 }
+
+#[cfg(test)]
```
Comment (sanpii) [inline src/task.rs]: I already wrote some tests in `tests/task.rs`. I dislike long file mixing code & tests.

File: src/task.rs [hunk 7]
```diff
@@ -84,3 +85,85 @@ impl ::std::fmt::Display for Task
...
+            "threshold_date": null,
+            "due_date": null,
+            "projects": [],
+            "hashtags": [ "tag_a", "tag_b" ],
+            "tags": {}
+        }"#;
+
+        let task : Task = ::serde_json::from_str(json).unwrap();
+
+        assert_eq!(task.subject, "Test");
```
Comment (sanpii) [inline src/task.rs]: I wonder if it’s not more readable to have only one assert, something likes:
```rust
let expected = ::task::Task {
    …
};

assert_eq!(expected, actual);
```

File: src/task.rs [hunk 8]
```diff
@@ -84,3 +85,85 @@ impl ::std::fmt::Display for Task
...
+        assert_eq!(task.due_date, Some(::Date::from_ymd(2018, 03, 03)));
+        assert!(task.projects.is_empty());
+        assert_eq!(task.hashtags[0], "tag_a");
+        assert_eq!(task.hashtags[1], "tag_b");
+        assert!(task.tags.is_empty());
+    }
+
+    #[test]
+    fn test_serialize_simple() {
+        let t    = Task::default();
```
Comment (sanpii) [inline src/task.rs]: `task` instead of `t` automatically justify the equal sign :yum:

File: Cargo.toml (hunk 3 above)
Comment (sanpii) [reply Cargo.toml]: We need to test both configurations. You can add a second test command in `.travis.yml` (in a perfect world `cargo test --all-features` covert all features but clippy probably fails).

**Comment by sanpii** [comment]
@matthiasbeyer Do you need some help to finish with PR?

**Comment by matthiasbeyer** [comment]
I am currently on a sabatical in noth america until early 2019. So I won't have time to finish this, nor the resources. I am sorry. I will have a look in early 2019 if that's okay, otherwise feel free to close or handle this in any way you think is appropriate.

**Comment by sanpii** [comment]
Merged.

**Pull request closed without merging**